src/DependencyInjection/DI/perf/ConstrainedGenericBenchmark.cs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using BenchmarkDotNet.Attributes;
+using Microsoft.Extensions.DependencyInjection.ServiceLookup;
+
+namespace Microsoft.Extensions.DependencyInjection.Performance
+{
+    public class ConstrainedGenericBenchmark
+    {
+        private static ServiceDescriptor[] _noConstraintSds;
+        private static ServiceDescriptor[] _newConstraintSds;
+        private static ServiceDescriptor[] _notMatchingNewConstraintSds;
+        private static ServiceDescriptor[] _simpleTypeConstraintSds;
+        private static ServiceDescriptor[] _complexInterfaceConstraintSds;
+        private static ServiceDescriptor[] _complexNotMatchingInterfaceConstraintSds;
+
+        [GlobalSetup(Target = nameof(NoConstraints))]
+        public void SetupNoConstraints()
+        {
+            _noConstraintSds = new[]
+            {
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
+            };
+        }
+
+        [Benchmark(Baseline = true)]
+        public void NoConstraints()
+        {
+            var csf = new CallSiteFactory(_noConstraintSds);
+            var _ = csf.GetCallSite(typeof(IEnumerable<IFakeGeneric<M>>), new CallSiteChain());
+        }
+
+        [GlobalSetup(Target = nameof(NewConstraint))]
+        public void SetupNewConstraint()
+        {
+            _newConstraintSds = new[]
+            {
+                new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>), ServiceLifetime.Transient),
```

Comment:
Make all of the implementation types be the same. Having only one of them being different downplays perf difference.

Reply:
Yes, it does after running the benchmark after. Fixed, will update benchmarks after the exception change.